compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -1186,6 +1186,13 @@ struct RootCollector<'a, 'tcx> {
 
 impl<'v> RootCollector<'_, 'v> {
     fn process_item(&mut self, id: hir::ItemId) {
+        let def_id = id.owner_id.def_id;
+        let (live_symbols, _) = self.tcx.live_symbols_and_ignored_derived_traits(());
+        if !live_symbols.contains(&def_id) && !self.tcx.sess.link_dead_code() {
+            // This is dead code; ignore it.
+            return;
+        }
```

Comment:
Why would this condition even happen when not compiling in incr comp mode? Also won't this remove all global_asm!() usages?

Reply:
> Why would this condition even happen when not compiling in incr comp mode?

I've observed this to happen with every symbol. https://github.com/rust-lang/rust/blob/d64517d6925e4824129fb952280bec2585166324/compiler/rustc_monomorphize/src/collector.rs#L379 looks at every item in the crate regardless of whether it's reachable.

I hadn't considered global_asm, I can special-case it here if you think it's necessary.

Reply:
Global_asm mut never be removed. Also please make this early exit conditional on the lazy collection mode. The eager collection mode doesn't omit anything on purpose to improve incr comp object file reuse.